cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -327,16 +327,22 @@ A function SHOULD preserve comments when input serialization format is YAML.
 This allows for human authoring of configuration to coexist with changes made by
 functions.
 
-### Annotations
-
-The orchestrator annotates resources in the wire format with annotation prefix
-`config.kubernetes.io`. These annotations are not persisted when the
-orchestrator writes the resources to the filesystem. The orchestrator sets this
-annotation when reading files from the local filesystem and removes the
-annotation when writing the output of functions back to the filesystem.
-
-In general, a function MUST NOT modify these annotations except the ones
-explicitly listed below.
+### Control Annotations
+
+For orchestration purposes, the orchestrator will use a set of annotations,
+referred to as _control annotations_, on resources in `Resources.items`. These
+annotations are not persisted to resource manifests on the filesystem: The
+orchestrator sets this annotation when reading files from the local filesystem
+and removes the annotation when writing the output of functions back to the
+filesystem.
+
+Annotation prefix `fn-ctrl.config.kubernetes.io` is reserved for use for control
+annotations. In general, a function MUST NOT modify these annotations. This
+enables orchestrator to add additional control annotations, without requiring
+changes to existing functions.
+
+For legacy reasons, a few control annotations do not use this annotation prefix.
```

Comment:
So what I propose:

1) Existing control annotations with prefix `config.kubernetes.io` (i.e. `index` and `path`) annotations remain as is.
2) `config.k8s.io/id` which is an anomaly should be renamed to `config.kubernetes.io/id`. As discussed this should be easy given no function should be consuming or mutating git.
3) Reserving `fn-ctrl.config.kubernetes.io` for "control annotations" to distinguish them from other annotations (e.g. `local-config`). Note that unlike annotations in 1), the language here is "MUST NOT" instead of "SHOULD NOT" to be conservative, but we can relax this.

I think there's agreement on Item 1 and 2 based on discussion in SIG-CLI discussion. Item 3 is debatable.

I agree there is complexity/inconsistency in definition of control annotations due to the need to remain backwards compatible. Function implementers need to know that control annotations are defined as all annotations with prefix `fn-ctrl.config.kubernetes.io` plus these legacy annotations listed here (`config.kubernetes.io/index`, `config.kubernetes.io/path`, `config.kubernetes.io/id`). Function libraries like kyaml can mitigate this complexity by encapsulating the complexity in a utility method.

Dropping the legacy control annotations would require a v2 of the API.

For reference, annotations used today in kustomize:

```
kustomize.config.k8s.io/id
kustomize.config.k8s.io/needs-hash
kustomize.config.k8s.io/behavior
kustomize.config.k8s.io/Inventory
kustomize.config.k8s.io/InventoryHash
kustomize.config.k8s.io/inventory-id

(Anomaly) config.k8s.io/id

config.kubernetes.io/path
config.kubernetes.io/index
config.kubernetes.io/function
config.kubernetes.io/previousNames
config.kubernetes.io/previousKinds
config.kubernetes.io/previousNamespaces
config.kubernetes.io/previousNames
config.kubernetes.io/prefixes
config.kubernetes.io/suffices
config.kubernetes.io/merge-source
config.kubernetes.io/local-config
```

WDYT?

Reply:
IMO if we reserve a new prefix, we should apply it to the legacy annotations too and temporarily emit both so tools can migrate from only looking for the legacy style (now) to only looking at the new style (future). I would rather have the less clear prefix permanently than a long-term situation where as you say tools have to be aware of both.

> Dropping the legacy control annotations would require a v2 of the API.

Which API? They don't actually appear in ResourceList. kyaml is still zero-series and so could do the aforementioned deprecation process before v1.

Re: the list used today in Kustomize, the ones using `kustomize.config.k8s.io` look Kustomize-internal and seem like they'd fall into the category of MUST be ignored by other tools. I'm not sure about some of the `config.kubernetes.io`. E.g. `config.kubernetes.io/merge-source` sounds internal to me, but it isn't really documented. Perhaps that could be part of the guidance--anything using prefix X (config.kubernetes.io and/or the new one we're discussing) should have a standard, documented meaning / value type. If it doesn't, a tool-specific prefix is a better choice.

Reply:
So these are the two options:

Option 1: The current wording. We have `internal` annotation prefix for future annotations, but do not touch `path` or  `prefix`.

Option 2: We standardize on `internal` annotation. But we should be aware that this IS a backwards incompatible change. Here are the migration steps using the `path` annotation as a concrete example:

1) Update all orchestrators so they add both legacy and new annotation when reading files. A function may mutate the legacy annotation (existing functions) or the new annotation (future functions). When writing files, the orchestrator should consume both and reconcile things correctly.
2) Update all functions so that they use the new annotation.
3) Orchestrator drops support for legacy annotation.

> Which API? They don't actually appear in ResourceList. kyaml is still zero-series and so could do the aforementioned deprecation process before v1.

The annotations are part of the spec which is independent from any implementation. When we say `kyaml` it refers two things:

a) __an implementation__ of a KRM function orchestrator.
b) __an implementation__ of a framework for developing a KRM function. 

b) is more relevant to this discussion so I'll focus on that. Not all functions use the `kyaml` library. Functions can be written using other frameworks (e.g. there's already a Typescript and Starlark one), or it can be developed using completely custom tooling. Having this standard is to enable a plethora of options for implementing functions. As a result, the version of the implementation is not important/relevant when it comes to backwards compatibility of the spec.

The issue is we don't have control to migrate all functions in step 2 above. There are functions in the wild that may not use kyaml (of if they are may not wish to update).

Having said that, I'm not necessarily opposed to option 2. It's trading a cleaner and more elegant spec for complexity of migration and some chance of breaking functions. I think these are mitigating factors that keeps the chance of breaking functions relatively low:

i) Only functions that are mutating internal annotations will be affected. The spec says functions "SHOULD NOT" change these. Although this is not a ban, the functions affected will be the exception rather than the rule.
ii) The functions affected most likely will be either in kustomize or kpt projects, which we can easily change.

So to summarize, I'm open to option 2 given we agree/appreciate a) it's a breaking change and b) Understand the complexity of migration.

Reply:
> IMO if we reserve a new prefix, we should apply it to the legacy annotations too and temporarily emit both so tools can migrate from only looking for the legacy style (now) to only looking at the new style (future). I would rather have the less clear prefix permanently than a long-term situation where as you say tools have to be aware of both.

I'm inclined to agree with this and have a slight preference for option 2 over option 1. I would prefer consistent prefixes, and if we don't clear it up now it will be much harder to fix later on. 

FWIW, these annotations:

```
config.kubernetes.io/previousNames
config.kubernetes.io/previousKinds
config.kubernetes.io/previousNamespaces
config.kubernetes.io/previousNames
config.kubernetes.io/prefixes
config.kubernetes.io/suffixes
config.kubernetes.io/allowNameChange
config.kubernetes.io/allowKindChange
```

are internal to kustomize, recently added, and I would be very surprised if any other tools are aware of them, so they can easily be changed based on guidance.

Reply:
Thanks everyone. Let's go with option 2. I filed an issue capturing all the changes we need in kyaml: https://github.com/kubernetes-sigs/kustomize/issues/4024